packages/block-editor/src/hooks/custom-fields.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { PanelBody, TextControl } from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { useBlockEditingMode } from '../components/block-editing-mode';
+
+/**
+ * Filters registered block settings, extending attributes to include `connections`.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	if ( hasBlockSupport( settings, 'connections', true ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			connections: {
+				type: 'object',
+			},
+		};
+	}
+
+	return settings;
+}
+
+/**
+ * Override the default edit UI to include a new block inspector control for
+ * assigning a connection to blocks that has support for connections.
+ * Currently, only the `core/paragraph` block is supported and there is only a relation
+ * between paragraph content and a custom field.
+ *
+ * @param {WPComponent} BlockEdit Original component.
+ *
+ * @return {WPComponent} Wrapped component.
+ */
+export const withInspectorControl = createHigherOrderComponent(
+	( BlockEdit ) => {
+		return ( props ) => {
+			const blockEditingMode = useBlockEditingMode();
+			const hasCustomFieldsSupport = hasBlockSupport(
+				props.name,
+				'connections',
+				false
+			);
+			// We prevent that the content is lost when the user removes the custom field.
+			// Editing the content in the paragraph block with a placeholder is not the best solution.
+			const prevContent = useRef( props.attributes?.content );
+			if ( ! prevContent.current ) {
+				prevContent.current = '';
+			}
+			if ( hasCustomFieldsSupport && props.isSelected ) {
+				return (
+					<>
+						<BlockEdit { ...props } />
+						{ blockEditingMode === 'default' && (
+							<InspectorControls>
+								<PanelBody
+									title={ __( 'Connections' ) }
+									initialOpen={ true }
+								>
+									<TextControl
+										__nextHasNoMarginBottom
+										autoComplete="off"
+										label={ __( 'Custom field meta_key' ) }
+										value={
+											props.attributes?.connections
+												?.attributes?.content?.value ||
+											''
+										}
+										onChange={ ( nextValue ) => {
+											if ( nextValue === '' ) {
+												props.setAttributes( {
+													connections: undefined,
+													content:
+														prevContent.current !==
+														''
+															? prevContent.current
+															: undefined,
```

Comment:
I'm not sure if we should do this because it causes another problem: 

If a paragraph was already connected to a custom field when you open the editor, then this will not do what we expect. When you clear the input field, the content gets replaced with the `This content will be replaced in the frontend by the custom field “<first letter of the field>” value.` placeholder.


https://github.com/WordPress/gutenberg/assets/5417266/a3076b18-f27e-453d-a4e6-bd03095c9774


I think that it's probably fine if we just set the content to to empty string once the input field is cleared.

Reply:
I agree, but also, in that case, if you had a very long paragraph written, you would lose the content.

Maybe we could create another paragraph if the meta_value goes from empty to a content. What do you think about that?

Reply:
> I think that it's probably fine if we just set the content to to empty string once the input field is cleared.
I will do it with placeholders to prevent adding content to the HTML p tag.